app/src/main/java/journal/gratitude/com/gratitudejournal/ui/security/AppLockFragment.kt

```diff
@@ -0,0 +1,104 @@
+
+class AppLockFragment : Fragment() {
+
+    private var fingerprintLock: Boolean = false
+
+    override fun onCreateView(
+        inflater: LayoutInflater, container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        return View(context)
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        val sharedPref = PreferenceManager.getDefaultSharedPreferences(context)
+        fingerprintLock = sharedPref.getBoolean("fingerprint_lock", false)
+        if (!fingerprintLock)
+            moveToTimeline()
+    }
+
+    override fun onResume() {
+        super.onResume()
+
+        if (fingerprintLock)
+            showFingerprintLock()
+    }
+
+    private fun showFingerprintLock() {
+        val executor = ContextCompat.getMainExecutor(context)
+        val biometricPrompt = BiometricPrompt(this, executor,
+            object : BiometricPrompt.AuthenticationCallback() {
+                override fun onAuthenticationError(
+                    errorCode: Int,
+                    errString: CharSequence
+                ) {
+                    super.onAuthenticationError(errorCode, errString)
+
+                    when(errorCode){
+                        BiometricConstants.ERROR_NEGATIVE_BUTTON, BiometricConstants.ERROR_USER_CANCELED ->{
+                            requireActivity().finish()
+                            return
+                        }
+                        BiometricConstants.ERROR_CANCELED ->{
+                            //happens when the sensor is not available
+                            //(happens onPause as well)
+                        }
+                        BiometricConstants.ERROR_NO_BIOMETRICS ->{
+                            //no finger print is setup
+                            Toast.makeText(
+                                context,
+                                "Please set up a fingerprint", Toast.LENGTH_SHORT
+                            ).show()
+                        }
+                        else ->{
+                            Toast.makeText(
+                                context,
+                                "Authentication error code $errorCode", Toast.LENGTH_SHORT
+                            ).show()
+                        }
+                    }
+                }
+
+                override fun onAuthenticationSucceeded(
+                    result: BiometricPrompt.AuthenticationResult
+                ) {
+                    super.onAuthenticationSucceeded(result)
+                    moveToTimeline()
+                }
+            })
+
+        val promptInfo = BiometricPrompt.PromptInfo.Builder()
+            .setTitle(getString(R.string.fingerprint_lock_title))
+            .setSubtitle(getString(R.string.fingerprint_lock_summary))
+            .setNegativeButtonText(getString(R.string.cancel))
+            .build()
```

Comment:
I think we can add this line so that the user doesn't have to additionally press confirm. I need to test with facial recognition to make sure though!
```suggestion
            .setConfirmationRequired(false)
            .build()
```

Reply:
There's also `.setDeviceCredentialAllowed(true)` which "allows you to specify whether it’s OK for devices that do not support biometric authentication to grant access through the device credentials, such as password, PIN, or pattern that are used to unlock the device" - [source](https://android-developers.googleblog.com/2019/10/one-biometric-api-over-all-android.html?m=1)

I haven't tried this out yet but thought it was worth mentioning. I'll update when I've tried it out.

Reply:
`setDeviceCredentialAllowed` sounds useful, would be nice to support simpler auth methods.
will try it out, and add if it's good